src/libraries/System.Security.Cryptography.X509Certificates/src/System/Security/Cryptography/X509Certificates/X509Certificate2.cs

```diff
@@ -628,6 +632,347 @@ public bool Verify()
+        ///   key with the label "ENCRYPTED PRIVATE KEY" is loaded. More advanced scenarios for loading certificates and
+        ///   private keys can leverage <see cref="System.Security.Cryptography.PemEncoding" /> to enumerate
+        ///   PEM-encoded values and apply any custom loading behavior.
+        /// </para>
+        /// <para>
+        /// For PEM-encoded keys without a password, use <see cref="CreateFromPem" />.
+        /// </para>
+        /// </remarks>
+        public static X509Certificate2 CreateFromEncryptedPem(ReadOnlySpan<char> certPem, ReadOnlySpan<char> keyPem, ReadOnlySpan<char> password)
+        {
+            using (X509Certificate2 certificate = ExtractCertificateFromPem(certPem))
+            {
+                string keyAlgorithm = certificate.GetKeyAlgorithm();
+
+                switch (keyAlgorithm)
+                {
+                    case Oids.Rsa:
+                        using (RSA key = ExtractKeyFromEncryptedPem<RSA>(keyPem, password, RSA.Create))
+                        {
+                            return certificate.CopyWithPrivateKey(key);
+                        }
+                    case Oids.EcPublicKey:
+                        using (ECDsa key = ExtractKeyFromEncryptedPem<ECDsa>(keyPem, password, ECDsa.Create))
+                        {
+                            return certificate.CopyWithPrivateKey(key);
+                        }
+                    case Oids.Dsa:
+                        using (DSA key = ExtractKeyFromEncryptedPem<DSA>(keyPem, password, DSA.Create))
+                        {
+                            return certificate.CopyWithPrivateKey(key);
+                        }
+                    default:
+                        throw new CryptographicException(SR.Format(SR.Cryptography_UnknownKeyAlgorithm, keyAlgorithm));
+                }
+            }
+        }
+
+        private static X509Certificate2 ExtractCertificateFromPem(ReadOnlySpan<char> certPem)
+        {
+            foreach ((ReadOnlySpan<char> contents, PemFields fields) in new PemEnumerator(certPem))
+            {
+                ReadOnlySpan<char> label = contents[fields.Label];
+
+                if (label.SequenceEqual(PemLabels.X509Certificate))
+                {
+                    // We verify below that every byte is written to.
+                    byte[] certBytes = GC.AllocateUninitializedArray<byte>(fields.DecodedDataLength);
+
+                    if (!Convert.TryFromBase64Chars(contents[fields.Base64Data], certBytes, out int bytesWritten)
+                        || bytesWritten != fields.DecodedDataLength)
+                    {
+                        // The contents should have already been validated by the PEM reader.
+                        throw new ArgumentException(SR.Cryptography_X509_NoPemCertificate, nameof(certPem));
+                    }
+
+                    return new X509Certificate2(certBytes);
```

Comment:
This will succeed at loading a cert from any supported file format that was sandwiched within `-----BEGIN CERTIFICATE-----`/`-----END CERTIFICATE-----`.

It's probably worth adding onto CertificatePal a method to only load the X.509 DER format.

For Windows the implementation is essentially

```C#
unsafe
{
    fixed (byte* spanPtr = span)
    {
        CRYPTOAPI_BLOB certBlob = new CRYPTOAPI_BLOB(
            span.Length,
            spanPtr);

        bool success = Interop.crypt32.CryptQueryObject(
            CertQueryObjectType.CERT_QUERY_OBJECT_BLOB,
            &certBlob,
            ExpectedContentTypeFlags.CERT_QUERY_CONTENT_FLAG_CERT,
            ExpectedFormatTypeFlags.CERT_QUERY_FORMAT_FLAG_BINARY,
            0,
            out _,
            out _,
            out _,
            IntPtr.Zero,
            IntPtr.Zero,
            out SafeCertContextHandle target);

        if (!success || target.IsInvalid)
        {
            target.Dispose();

            int hr = Marshal.GetHRForLastWin32Error();
            throw hr.ToCryptographicException();
        }

        return new CertificatePal(target);
    }
}
```

(which might require adding another overload to the CryptQueryObject P/Invokes)

Reply:
@bartonjs yeah I thought about that... something I had in there originally was attempting to crack open the decoded contents with `System.Security.Cryptography.X509Certificates.Asn1.CertificateAsn` and if decoding failed then "nope". Is that a reasonable approach or is a new method on the PAL loader preferred?

Reply:
I'm OK with forcing it through an X.509-shaped hole before calling the existing ctor.  Just looking to avoid supporting PEM "certificates" that are actually null/empty-password PFXes until the end of time.